tests/test_models.py

```diff
@@ -0,0 +1,886 @@
+
+    return probs, preds, labels, save_dir_path
+
+
+def _get_outputs_api3(dataset, backbone, nr_classes=9):
+    """Helper function to get the model output using API 3."""
+
+    # API 3
+    model = CNN_Patch_Model(backbone=backbone, nr_classes=nr_classes)
+
+    # coverage setter check
+    model.set_preproc_func(lambda x: x - 1)  # do this for coverage
+    assert model.get_preproc_func()(1) == 0
+    # coverage setter check
+    model.set_preproc_func(None)  # do this for coverage
+    assert model.get_preproc_func()(1) == 1
+
+    predictor = CNN_Patch_Predictor(model=model, batch_size=1, verbose=False)
+    # don't run test on GPU
+    output = predictor.predict(
+        dataset, return_probs=True, return_labels=True, on_gpu=False
+    )
+
+    probs = output["probs"]
+    preds = output["preds"]
+    labels = output["labels"]
+
+    return probs, preds, labels
+
+
+def test_create_backbone():
+    """Test for creating backbone."""
+    backbone_list = [
+        "alexnet",
+        "resnet18",
+        "resnet34",
+        "resnet50",
+        "resnet101",
+        "resnext50_32x4d",
+        "resnext101_32x8d",
+        "wide_resnet50_2",
+        "wide_resnet101_2",
+        "densenet121",
+        "densenet161",
+        "densenet169",
+        "densenet201",
+        "googlenet",
+        "mobilenet_v2",
+        "mobilenet_v3_large",
+        "mobilenet_v3_small",
+    ]
+    for backbone in backbone_list:
+        try:
+            get_model(backbone, pretrained=False)
+        except ValueError:
+            assert False, "Model %s failed." % backbone
+
+    # test for model not defined
+    with pytest.raises(ValueError, match=r".*not supported.*"):
+        get_model("secret_model", pretrained=False)
+
+
+def test_predictor_crash():
+    """Test for crash when making predictor."""
+    # test abc
+    with pytest.raises(NotImplementedError):
+        Model_Base()
+    with pytest.raises(NotImplementedError):
+        Model_Base.infer_batch(1, 2, 3)
+
+    # without providing any model
+    with pytest.raises(ValueError, match=r"Must provide.*"):
+        CNN_Patch_Predictor()
+
+    # provide wrong unknown predefined model
+    with pytest.raises(ValueError, match=r"Predefined .* does not exist"):
+        CNN_Patch_Predictor(predefined_model="secret_model")
+
+    # provide wrong model of unknown type, deprecated later with type hint
+    with pytest.raises(ValueError, match=r".*must be a string.*"):
+        CNN_Patch_Predictor(predefined_model=123)
+
+    # model and dummy input
+    model = CNN_Patch_Predictor(predefined_model="resnet34-kather100K")
+    img_list = [
+        np.random.randint(0, 255, (4, 4, 3)),
+        np.random.randint(0, 255, (4, 4, 3)),
+    ]
+    # only receive a dataset object
+    with pytest.raises(ValueError, match=r".*torch.utils.data.Dataset.*"):
+        model.predict(img_list)
+
+
+def test_set_root_dir():
+    """Test for setting new root dir."""
+    # skipcq
```

Comment:
What does this mean?

Reply:
As in when user want to change `rcParam['TIATOOLBOX_HOME']` before downloading pretrained data. This is to test if the current setup allow us to modify the content consistently after subsequent re-import.

Reply:
In case you are asking about `# skipcq` this is to silence deep source from marking reimporting within this func as error because we need to do that to check.